docs/src/examples/howtorigidtransforms.md

```diff
@@ -0,0 +1,205 @@
+t1b, w1b = [1.0; 0],  get_vector(M_R, R0, π/2, B)
+
+# walk positive y and turn left again
+t2, w2 = [1.0; 0],  get_vector(M_R, R0, π/2, B)
+
+# walk negative x and turn left
+t3, w3 = [2.0; 0],  get_vector(M_R, R0, π/2, B)
+
+# walk negative y back onto origin and turn left
+t4, w4 = [1.0; 0],  get_vector(M_R, R0, π/2, B)
+```
+
+Notice that each of these coordinates are **not** defined against the origin, but are fully relative translations and rotations.
+
+Before we can compose these separate translation and rotation pairs, we need to leverage some pairing data structure by which Manifolds.jl can identify a Product manifold operation is needed.  There are currently two ways to do so.  The `compose` hereafter can work with either the `ProductRepr` or `ProductArray` representations.
+
+### ProductRepr for SpecialEuclidean(2)
+
+Let's first define the identity group element on the manifold directly:
+```julia
+p0 = ProductRepr(Tr0, R0)
+```
+
+!!! note
+    Using the language or reference frames, this is the identity reference to which other points will be defined.  For example, consider measuring various points relative to a common "reference frame", then `p0` might represent that reference point in-and-about the other points on the same `SpecialEuclidean(2)` manifold.
+
+Above, we defined individual tranlate and rotate vectors such as `t2,w2`.  The next step towards using `compose` for rigid body transformations is to properly pair (i.e. data structuring) the Lie algebra elements and then map to the equivalent Lie group element on the group manifold `G` or `M`.
+
+The trivial case is a zero vector from the point `p0`
+```julia
+# make Lie Algebra element 
+x0 = ProductRepr(t0, w0)
+
+p0_ = exp(G, p0, x0)  # or,  exp(M, p0, x0)
+```
+
+The exponential maps the Lie algebra to the associated Lie group, but in this trivial zero-vector case should be exactly the same point on the manifold
+```julia
+@assert isapprox(p0_.parts[1], p0.parts[1]); @assert isapprox(p0_.parts[2], p0.parts[2])
+```
+
+All the other relative vector segments are mapped onto the manifold, and carefully not relative to the point `p0` in each case -- we will soon show why
+```julia
+# calculate the exponential mapping from point p0
+x1a = ProductRepr(t1a, w1a)   # Lie algebra element
+p1a = exp(G, p0, x1a)         # Lie group element
+
+x1b = ProductRepr(t1b, w1b)
+p1b = exp(G, p0, x1b)
+
+x2 = ProductRepr(t2, w2)
+p2 = exp(G, p0, x2)
+
+x3 = ProductRepr(t3, w3)
+p3 = exp(G, p0, x3)
+
+x4 = ProductRepr(t4, w4)
+p4 = exp(G, p0, x4)
+```
+
+If you want to continue using `ProductRepr` then skip ahead to [Compose Rigid Transforms, 2D](@ref compose_rigid_transforms_2d).
+
+### ProductArray for SpecialEuclidean(2)
+
+As mentioned earlier, there are two data representation formats available for product manifolds.  This paragraph will repeat the same steps as the `ProductRepr` paragraph above, but using the `ProductArray` representation instead.
```

Comment:
I've done some work on a new power manifold representation which should make `ProductArray` essentially obsolete: https://github.com/JuliaManifolds/Manifolds.jl/pull/367 . What do you think? That new representation should offer comparable performance and be easier to use, at the price of having to implement non-mutating variants of some operations on manifolds separately from the mutating ones.

Reply:
We can definitely change the text on this page to whatever is best for Manifolds.jl and to keep the math rigor / utility.  I'm 100% good with that :-)  Thanks for taking the time to help!  I did find `ProductRepr` a little easier at first, since the `shape` definitions added another layer to learn on day one.

Perhaps it is worth saying what I struggled with in the beginning, which is why I drafted these pages as they are/were.  I was initially very confused trying to pick apart the differences between: 
- Group vs. `base_manifold( of the Group )`, which do I use because sometimes I get an error, other times not...
- `ProductRepr` vs `ProductArray`, are these the same, is one a coordinate, do they work the same on all Manifolds...
- What is a `ShapeSpecification`, and should I worry about `Static`...
- What's up with `hat` vs `get_vector`...
- is `vee` (where textbooks say 'vectorize') the same as `get_vector`...
- does `get_coordinates` take either group or vector (i.e. algebra) elements, and should I always write that myself...
- are 'coordinates' the same thing as a 'vector' and what does that have to do with `shape`, and how much does Julia's own ::Array get used for dispatch...
- if you call `exp`, does the code know whether you are giving it coordinates or a algebra element...
- What does Manifolds.jl even define as 'coordinates', I guess it has something to do with the `DefaultBasis`, maybe...

So hopefully these two new tutorials will help newcomers breeze through these (perhaps silly first order) questions, and quickly get to the harder stuff like "how do we implement ad PR a new dispatch for this or that".

The reason I listed both representations is simply to help others get through the learning curve faster, and help contribute the commits needed towards a consolidated solution.  Until a clear design template and "Julian" philosophy about which representation is "better" emerges, I think it will be difficult to say option 'A' vs. option '1' is better.  The wide open aperture about data types in the representations (coordinates, vectors, groups) makes this a little harder to pin down in the early days.

I don't yet know the implementation details well enough to see how / where the differences between `ProductRepr` and `ProductArray` manifest.  I might be a bit more pragmatic in this sense and happy to work with what's available without obstructing progress.  In utopia, yes only a single one-size-fits-all representation would be nice.  I hope to help contribute some code as I learn more, but that is a longer term thing.

So my suggestion then is perhaps we should move the `ProductArray` paragraph of this tutorial right to the very end of the page, and simply point out "please see the end of this page for an earlier alternative `ProductArray` data representation".  The `ProductArray` representation is likely already in use.  My feeling is the community will ultimately adopt one approach over the other, and most importantly the rigor/utility should not be lost.  Good design and clean-up will result in fast code, but it has to be human readable so that a sizable audience can share the burdens.  I think the risk now is the community adopts the "easier" vs the "better" approach.  It's probably too soon to try force a deprecation any which way.

Reply:
> Thanks for taking the time to help!

You're welcome :slightly_smiling_face: . I'm happy that you find Manifolds.jl useful.

> I did find `ProductRepr` a little easier at first, since the `shape` definitions added another layer to learn on day one.

I'm really not satisfied with the design `ProductArray`, and I'm almost sure that we can make `ProductRepr` cover all possible use cases for product manifolds, so I'd suggest just reporting problems with that approach and not using `ProductArray` at all.

> * Group vs. `base_manifold( of the Group )`, which do I use because sometimes I get an error, other times not...

Calling `base_manifold` should rarely be necessary. Feel free to report any cases where calling an operation on `base_manifold` works but directly on a group it does not.

> * `ProductRepr` vs `ProductArray`, are these the same, is one a coordinate, do they work the same on all Manifolds...

I think you should just stop using `ProductArray`. It was an idea I had when I doubted `ProductRepr` could work with broadcasting and power manifolds, but now `ProductRepr` does work there.

> * What is a `ShapeSpecification`, and should I worry about `Static`...

Same as above, you shouldn't use `ShapeSpecification`.

> * What's up with `hat` vs `get_vector`...
> * is `vee` (where textbooks say 'vectorize') the same as `get_vector`...

`vee` and `hat` were supposed to be aliases to certain variants of `get_coordinates` and `get_vector` to make people familiar with them more comfortable but it doesn't seem to work this way... Personally, I never use `hat` and `vee`.

> * does `get_coordinates` take either group or vector (i.e. algebra) elements, and should I always write that myself...

`get_coordinates` actually takes both a group element and an algebra element. For groups where tangent vectors are represented as tangents at the identity element of the group, it does not matter what point you give there. But note that `get_coordinates` originates in differential geometry, not Lie theory, and as such it needs to know which point the vector is tangent at.

> * are 'coordinates' the same thing as a 'vector' and what does that have to do with `shape`, and how much does Julia's own ::Array get used for dispatch...

I don't quite understand the first part of the question, could you rephrase it? I guess it might be confusing that `get_coordinates` returns a `::Vector` (or, sometimes `SVector`), but the thing you get there is the vector of coordinates, while `get_vector` returns a tangent vector (or a vector from a fibre of some other vector bundle but maybe let's leave that for another time -- I only mention it to explain why it isn't called `get_tangent_vector`).

I think we almost never dispatch on Julia's `Array`, though in many cases we dispatch on `AbstractArray`.

> * if you call `exp`, does the code know whether you are giving it coordinates or a algebra element...

`exp` always expects an element of the tangent space (so an algebra element), it may even silently give a wrong result if given coordinates. We have, though, `ValidationManifold` for verifying that you give correct arguments to functions in your algorithms.

> * What does Manifolds.jl even define as 'coordinates', I guess it has something to do with the `DefaultBasis`, maybe...

This might be a decent explanation (by @kellertuer :slightly_smiling_face: ): https://www.youtube.com/watch?v=md-FnDGCh9M&t=1060s . 'Coordinates' in Manifolds.jl generally refer to coordinates of a tangent vector (or f.e. cotangent vector also) in a basis, like in basic linear algebra: https://www.math.tamu.edu/~yvorobet/MATH304-2011A/Lect2-07web.pdf . The catch here is, while you can do lots of linear algebra ignoring the (often stated without appropriate explanation) distinction between elements of vector spaces and their coordinates with respect to a basis, the distinction is absolutely necessary in differential geometry. In Manifolds.jl we usually store tangent vectors as numbers that do not correspond to coefficient in any basis. This probably still needs some better explanation in docs.

> So hopefully these two new tutorials will help newcomers breeze through these (perhaps silly first order) questions,

I think these questions are quite good actually, and show the deficiencies in our documentation. I'm not sure, though, where to put some of these answers, for example these regarding coordinates.

> The reason I listed both representations is simply to help others get through the learning curve faster, and help contribute the commits needed towards a consolidated solution. Until a clear design template and "Julian" philosophy about which representation is "better" emerges, I think it will be difficult to say option 'A' vs. option '1' is better. The wide open aperture about data types in the representations (coordinates, vectors, groups) makes this a little harder to pin down in the early days.
> 
> I don't yet know the implementation details well enough to see how / where the differences between `ProductRepr` and `ProductArray` manifest. I might be a bit more pragmatic in this sense and happy to work with what's available without obstructing progress. In utopia, yes only a single one-size-fits-all representation would be nice. I hope to help contribute some code as I learn more, but that is a longer term thing.
> 
> So my suggestion then is perhaps we should move the `ProductArray` paragraph of this tutorial right to the very end of the page, and simply point out "please see the end of this page for an earlier alternative `ProductArray` data representation". The `ProductArray` representation is likely already in use. My feeling is the community will ultimately adopt one approach over the other, and most importantly the rigor/utility should not be lost. Good design and clean-up will result in fast code, but it has to be human readable so that a sizable audience can share the burdens. I think the risk now is the community adopts the "easier" vs the "better" approach. It's probably too soon to try force a deprecation any which way.

I'm currently 99% sure `ProductArray` should be just removed, and maybe I should've pulled the plug earlier. One particular breakthrough in this area is quite recent but I still feel a little embarrassed that you had to go through `ProductArray`. `ProductRepr` is definitely easier, and with the recent changes I'm certain it can also always be as fast, if not faster, than `ProductArray`.

The one thing that remains to be done is changing "can be" to "is" in the last sentence, which requires subverting a little our approach to non-mutating functions -- specifically, we will need to explicitly write some non-mutating methods instead of relying on allocation and mutating variants. I'll write more about it separately, the key thing here is that you should report any performance issues that may be caused by Manifolds.jl -- I'm aware of a lot of places where performance could be improved but I have limited time so I can't fix every potential issue :slightly_smiling_face: .

Finally, thank you for your comments! Also, feel to ask them more often, after reading them it looks like you haven't been asking enough questions when working through Manifolds.jl :slightly_smiling_face: . In any case I and Ronny are available at Julia's slack and zulip if you want some real-time conversation.

Reply:
Uh! Neat, that even my talk from last year was linked; now you now how I sound and how much German accent my English has ;)

I looked a bit at `vee` and `hat` (cf. https://en.wikipedia.org/wiki/3D_rotation_group#Lie_algebra) and it seems that there is one confusion with your question

> * is vee (where textbooks say 'vectorize') the same as get_vector...

First of all – as far as I found hat/vee are used only with rotations and not even with Lie groups in general.

If I understand the wikipedia article corretcly; the `hat` function takes a set of coordinates $\omega$ and returns an alement from the Lie algebra. So with the identification of every tangent space `hat` reconstructs a vector from given coefficients in a basis, so `hat`=`get_vector` (to be precise https://github.com/JuliaManifolds/ManifoldsBase.jl/blob/194e27e985292e5e1685cf10e7bf27d4af40b9f7/src/bases.jl#L788).
On the other hand, `vee`(cf also https://juliamanifolds.github.io/Manifolds.jl/stable/interface.html#ManifoldsBase.vee-Tuple{AbstractManifold,Any,Any}) maps a vector that is given in a certain basis (note the Einstein summation on the left) to the coefficients, so it is `vee`=`get_coordinates` for the special case of using the `VeeOrthogonalBasis` (this is also the basis for `hat`s `get_vector` by the way) … see the one code line https://github.com/JuliaManifolds/ManifoldsBase.jl/blob/194e27e985292e5e1685cf10e7bf27d4af40b9f7/src/bases.jl#L910).
We should maybe be more precise in this description in vee/hat, sure. My main reason for not having done that yet is, that I never used them until now.

Similarly to

> * does get_coordinates take either group or vector (i.e. algebra) elements, and should I always write that myself...

it takes a tangent vector (Lie algebra element) and returns coefficients (coordinates wrt that basis).

Reply:
> Calling base_manifold should rarely be necessary. Feel free to report any cases where calling an operation on base_manifold works but directly on a group it does not.

Ah, that helps thanks!

> vee and hat were supposed to be aliases to certain variants of get_coordinates and get_vector to make people familiar with them more comfortable but it doesn't seem to work this way... Personally, I never use hat and vee.

Thanks, kellertuer also mentioned that.  Hopefully 355 and 366 help newcomers through that faster than I was able to.

> get_coordinates actually takes both a group element and an algebra element. For groups where tangent vectors are represented as tangents at the identity element of the group, it does not matter what point you give there. But note that get_coordinates originates in differential geometry, not Lie theory, and as such it needs to know which point the vector is tangent at

Is this written down anywhere in the current docs, I think that should be linked from these tutorial pages -- this is a good piece of info, thanks!

> I don't quite understand the first part of the question, could you rephrase it? I guess it might be confusing that get_coordinates returns a ::Vector (or, sometimes SVector), but the thing you get there is the vector of coordinates, while get_vector returns a tangent vector (or a vector from a fibre of some other vector bundle but maybe let's leave that for another time -- I only mention it to explain why it isn't called get_tangent_vector).

Oh all good, that bullet point list above is just a quick list of problems I was not able to easiliy resolve from the documentation when I started.  Hence those bullet points are specific things I'm trying to answer with these two new tutorials 355 and 366.

> We have, though, ValidationManifold for verifying that you give correct arguments to functions in your algorithms.

- [ ] Ah, should add that in these two new pages somewhere too.

> This might be a decent explanation (by kellertuer slightly_smiling_face ):

- [ ] ha, should definitely link that in these docs too.  I'm taking this as an opportunity to boil down and capture all the conversations we've had in the past months into the two new shorter tutorials.

> I think these questions are quite good actually, and show the deficiencies in our documentation. I'm not sure, though, where to put some of these answers, for example these regarding coordinates.

Great, great, let me chew on where to fit more of these details into the tutorials.  I think most of this latest volley of notes should either be linked out to elsewhere at the right place.  It should also be okay to add a paragraph or two for additional details.

I don't think these two new "tutorials" (355 and 366) are all thats needed at this stage.  They are specifically about the end-to-end usage of two well known manifolds for newcomers at varying levels of prior knowledge.  So the documentation criteria for me on these two pages is to rapidly learn the conventions, with basic do's and don'ts.  The additional step then is a reader will either want verification of details if they are an expert (i.e. links to particular functions) and which headings are actually listed here.  Another reader might first want to figure out the relation between for e.g. `exp` and `retraction` (as we spoke through last week).

> I'm currently 99% sure ProductArray should be just removed, and maybe I should've pulled the plug earlier. One particular breakthrough in this area is quite recent but I still feel a little embarrassed that you had to go through ProductArray. ProductRepr is definitely easier, and with the recent changes I'm certain it can also always be as fast, if not faster, than ProductArray.

oh all good, Manifolds.jl is awesome!  You guys have been a great help!  Just moving to `ProductRepr` sounds good to me, and then focus effort on building out the required dispatches.  I don't know enough of the internals just yet, but I'm slowly getting there.

>  you should report any performance issues that may be caused by Manifolds.jl -- I'm aware of a lot of places where performance could be improved but I have limited time so I can't fix every potential issue slightly_smiling_face .

All good thanks, we are likely to stumble into a few and have many performance related things to resolve too.  Will help however best I can as things come up, at least just document.

> If I understand the wikipedia article corretcly; the hat function takes a set of coordinates $\omega$ and returns an alement from the Lie algebra. So with the identification of every tangent space hat reconstructs a vector from given coefficients in a basis, so hat=get_vector

That's how I understand it now too.  But not at all how I understood things at first -- I was trudging through runtests.jl to try understand how to do the end-to-end usage.  It's much clearer now :-) And the problem I think is really just that first usage example, rather than a listing of function features.  It's fine if you know which functions you want, but the problem is which one of the 100's is it... etc.  So hopefully 355 and 366 helps.